### PR TITLE
Remove error thrown by joinpath (fixes #20912)

### LINKED ISSUE #20912: joinpath should not throw. period.
As observed in https://github.com/JuliaLang/julia/issues/20695, `joinpath` causes an error on Windows in situations where a git URL looks like a relative path with a drive specification – since the drives of the two arguments don't match, an error is thrown. That specific problem can be worked around, but the deeper problem is that `joinpath` throws in the first place. It's fundamentally just a string manipulation function and shouldn't throw errors. However, that leaves the question of what should happen when the second argument to `joinpath` is a relative with a drive that doesn't match that of the first argument.

### PR DESCRIPTION
**Pull request by SaschaMann**
`joinpath` now returns the second path instead of throwing an `ArgumentError` if the drives of the two arguments do not match. ([#20912]).

### CODE REVIEW

File: NEWS.md
Comment (StefanKarpinski) [inline NEWS.md]: Might be worth a parenthetical "(Windows only)".

**Comment by StefanKarpinski** [comment]
Can someone else who understands how Windows paths work take a look at the tests? cc @vtjnash, @tkelman

**Review by StefanKarpinski — APPROVED** [review]

**Comment by SaschaMann** [comment]
Added the "Windows only" and some info to the docstring about this behaviour.

File: base/path.jl
Comment (tkelman) [inline base/path.jl]: second, or last?

Comment (SaschaMann) [inline base/path.jl]: The last. Fixed the docstring.

**Comment by SaschaMann** [comment]
I'll squash these once it's ready to merge.

**Comment by tkelman** [comment]
hm it's not quite the last argument right, it's the last argument that starts with a drive letter? similar to the absolute path behavior everywhere else, and also true when drives do match?

**Comment by SaschaMann** [comment]
Yea, it's not the last argument but the last absolute path, eg

```julia
julia> joinpath("A:", "a", "B:", "b", "C:", "c")
"C:c"
```

**Comment by StefanKarpinski** [comment]
It might be easier to explain the two-argument version and then say that this is called recursively.

**Comment by StefanKarpinski** [comment]
I still don't think this accurately describes the behavior 😞

**Comment by SaschaMann** [comment]
> It might be easier to explain the two-argument version and then say that this is called recursively.

So splitting up the docstring in two parts, one for `joinpath(a::String, b::String)` (containing info about the standard behaviour and the part that's added in this PR), and the other one for `joinpath(parts...)` (refering to the first one and the recursion)?

**Comment by StefanKarpinski** [comment]
I'm also pretty sure this definition is buggy – my fault since my original patch was wrong. It should return `b` not `B`. This is why the tests looked fishy, which is why I asked for people familiar with Windows to take a look. Can set things so I can push to your branch? I have a rewording and a fix. Or maybe it's easier if I just make a new PR and squash all this.

**Comment by SaschaMann** [comment]
Feel free to push to this, I have edits from maintainers enabled.

**Comment by StefanKarpinski** [comment]
Ok, I tried that but it didn't work. Maybe my push incantation was wrong.

**Comment by StefanKarpinski** [comment]
I was trying to push via git:// protocol which doesn't work, of course.

**Comment by StefanKarpinski** [comment]
Crap, no, that's not right either.

**Comment by SaschaMann** [comment]
The tests are wrong since now the drive will be dropped if two drives don't align:

```julia
julia> joinpath(S("foo"),S("bar:baz"))
"baz"

julia> joinpath(S("C:"),S("foo"),S("D:"),S("bar"))
"bar"
```

**Comment by SaschaMann** [comment]
Yeah now it should work.

**Comment by StefanKarpinski** [comment]
Yeah, I pushed a re-fixed version and updated the tests. Hopefully this is correct now.

**Comment by SaschaMann** [comment]
Locally it works as expected 👍

**Comment by StefanKarpinski** [comment]
Rebased on latest master and force pushed since there was a broken merge in doc stuff.

File: test/path.jl
Comment (Sacha0) [inline test/path.jl]: Spaces between arguments might improve readability :).

**Review by Sacha0 — APPROVED** [review]
:) !

**Pull request merged**
